week4/preparation.md

```diff
@@ -1,16 +1,162 @@
+
+    // Loop over the list of ids to create a task for each
+    foreach (int userId in userIds)
+    {
+        // Same as above, but in one line. Create a new task and add it directly in the list
+        getUserTasks.Add(GetUserAsync(userId));
+    }
+
+    // While the list from above has tasks, loop over them
+    while (getUserTasks.Any())
+    {
+        // When any of the tasks is done assign it to the variable
+        var user = await Task.WhenAny(getUserTasks);
+
+        // We then have to remove the task from the list of tasks as it is completed
+        getUserTasks.Remove(user);
+
+        // Log to the console the retrieval of each user
+        System.Console.WriteLine($"User {user.Name} was retrieved");
+    }
+}
+```
+
+## Usefull following material:
+
+- [Task asynchronous programming model](https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming/task-asynchronous-programming-model)
+- [Asynchronous programming](https://learn.microsoft.com/en-gb/dotnet/csharp/asynchronous-programming/async-scenarios)
+- [Video: Introduction To Async, Await, And Tasks](https://www.youtube.com/watch?v=X9N5r6kMOxw)
+- [Async explained with common scenario](https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming)
+
+
+## Exercises
+1. __Get item by id__: Create a new endpoint based on the following
+    - is `GET`
+    - takes an `id` as a query parameter (`int`)
+    - executes an async `GET HTTP` call to an endpoint using the id from the query parameter
+    - checks if the response is successful
+    - returns the response or `Results.BadRequest` if the response fails
+
+> **_NOTE:_**
+>- a public api that returns dummy data can be found here: 
+    https://dummyjson.com/; e.g. https://dummyjson.com/products/1
+>- to check if an api call is successful the `HttpResponseMessage` has a property called `IsSuccessStatusCode` e.g. `response.IsSuccessStatusCode`
+>- to return an object from a response you can use 
+    `ReadFromJsonAsync(typeof(object))` instead of `ReadAsStringAsync()`
+
+2. __Get users with name__: Create a new endpoint based on the following:
+    - is `GET`
+    - takes a `name` as a query parameter 
+    - validates that the `name` is not null or an empty string (if it is it returns a `BadRequest`)
+    - calls (async) the same dummy endpoint and returns a list of users (id, firstName, lastName) that have been found
+> **_NOTE:_**
+>- the dummy page provides this endpoint that you can use: https://dummyjson.com/users/search?q=`searchquery`
+>- the response from the endpoint will be an object which will contain a list of users with a lot of data
+    >> - you only need to return the 3 properties specified above
+    >> - use a mix of `ReadFromJsonAsync` and define 2 new classes
+    >> - One would be the main `Response` (the response with a List of Users inside) 
+    >> - and the second would be the actual `User` class in which you can define the required properties.
```

Comment:
@tpetrina @Vertygo Do you think these are too difficult as preparation? 
From my pov I think they are doable as in the examples above I have specified how to do most of these and I am trying to guide them with the notes as well 🤔